[Woo POS] Abstract `POSItemProvider`/`POSItem` and move it to Yosemite

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12879

## Description
This PR tackles decoupling and abstraction between the POS and the origin of the data by moving the data providing logic to Yosemite, and defining initial interfaces for the POS items.

This is the first step so we can keep refining the model and details in future iterations.

## Changes:
- Defines `POSItemProvider` and `POSItem` protocols. 
- Moved models and data provider to Yosemite
- General renaming for all objects/methods/views to use `item` instead of `product`
- POS only knows about the interfaces

## Question
- I'm not quite sure if we're instantiating the item provider in the right place at the moment (`HubMenuViewModel`). Any suggestion if/where this should be moved?

## Testing instructions
- In POS mode, observe that products loaded in the grid are still products from your store.

iamgabrielma committed:
Abstract POSItemProvider. Move to Yosemite

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Merge branch 'trunk' into task/12879-pos-move-product-provider-to-yosemite

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/PointOfSaleEntryPointView.swift
#	WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift

iamgabrielma committed:
resolve merge conflict with rename

iamgabrielma committed:
linting

iamgabrielma committed:
Move POSItem models to their own files

iamgabrielma committed:
Move product provider to Yosemite

We need to disable previews temporarily due to the usage of service locator as dependency, since this is within the woocommerce binary

iamgabrielma committed:
remove currency dependency from product model

iamgabrielma committed:
Internal comments

joshheald commented on WooCommerce/Classes/POS/Presentation/ProductCardView.swift:
```diff
@@ -14,7 +15,7 @@ struct ProductCardView: View {
                 Text(product.name)
                     .foregroundStyle(Color.primaryBackground)
-                Text(product.priceWithCurrency)
+                Text(product.formattedPrice)
```
This is a little nitpicky, but one of the benefits of having really view-friendly models is that we can use clearer names.

Why not make this just `price`? We can specialise the name for other cases if we need them in future.

iamgabrielma replied:
Agree, this was just a renaming but I think we need to rework the current way currency is applied to the prices. We can also see some duplicity in the entry point, where the dependency is injected directly to the view but is not necessary since this can be located through the view model. I've added a note to tackle this separately.

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -26,10 +29,12 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
     private let currencyFormatter: CurrencyFormatter
 
-    init(products: [POSProduct],
+    init(products: [POSItem],
```
```suggestion
    init(items: [POSItem],
```

iamgabrielma replied:
Yep yep! I've renamed all `product` references to `items` in different commits

joshheald commented on WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:
```diff
@@ -26,10 +29,12 @@ final class PointOfSaleDashboardViewModel: ObservableObject {
-        self.products = products
+        // For the moment we inject POSItem, but only use the POSProduct implementation, since it's the only item type we're using
+        // Later we might change this to support other types of items
+        self.products = products.compactMap { $0 as? POSProduct }
```
Why not just use items from the beginning? 

Specialising here doesn't get us most of the benefits, I think we should be able to integrate with just the `POSItem` protocol quite easily, if the protocol is well designed.

iamgabrielma replied:
Just me failing to fight against SwiftUI type inference, by making the change you suggested [here](https://github.com/woocommerce/woocommerce-ios/pull/12882#discussion_r1621674601) this is resolved now 🎉

joshheald commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenu.swift:
```diff
@@ -162,7 +162,14 @@ private extension HubMenu {
+                                                                         siteID: ServiceLocator.stores.sessionManager.defaultSite?.siteID ?? 0,
+                                                                         currencySettings: ServiceLocator.currencySettings)
+
+                PointOfSaleEntryPointView(itemProvider: posProductProvider,
```
I'm being picky, but if we avoid thinking/coding expectations that these are products, we're more likely to do it right first time.

iamgabrielma replied:
Definitely, on https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/605fc4b2e53ee04bac8d7755df0483c9d54562bc we hide the implementation detail behind the protocol.

joshheald commented on Yosemite/Yosemite/PointOfSale/POSItemProvider.swift:
```diff
@@ -0,0 +1,7 @@
+public protocol POSItem {
+    var itemID: UUID { get }
+}
```
Design this to match the designs. If we get this right, there's no need to specialise to products when we're building the view.
```suggestion
public protocol POSItem {
    var itemID: UUID { get }
    var name: String { get }
    var price: String { get } // It's arguable whether this should be pre-formatted in Yosemite or done in the view.
    var imageURL: URL { get } // This is not an image, so we can lazy load the image in the view. There are probably other ways
    var details: [String] { get } // A very simple way to do category/subcategory strings. We'd probably grow out of it, and an array of enum cases might give more flexibility. We can still have specific named properties for these on the concrete implementations though, and make this a computed var which puts together the named properties in the right order.
    func makeCartItem() -> POSCartItem // Called to make something when it's time to add to the cart.
}
```

iamgabrielma replied:
Oh! I didn't know this. I used `itemID` just as an initial property to get the ball rolling with the protocol, and didn't realize that the miss-match here is what was causing the casting problem in SwiftUI. I updated the model here a2469ec73e2521e03ff58a96d87bd23539cb5611 so we can remove the downcast and will update the property names and adjust the model on different commits.

joshheald reviewed (commented):
Added some notes. 

The general point is that we should avoid downcasting to a concrete item type. If we downcast, we won't consider these as generic items when we're building the view, so we'll likely end up in the same place as we did with Products vs ProductVariations on the order creation screen.

I'll leave @jaclync to cover the ItemProvider as I couldn't find her comment/reasoning around it, so I'm not sure if you've covered the cases she was looking for.

iamgabrielma committed:
Update model to match design. Remove type downcast

iamgabrielma committed:
Products to Items refactor

iamgabrielma committed:
Rename views

iamgabrielma committed:
Restore previews

iamgabrielma committed:
Update ItemRowView name and refactor

iamgabrielma committed:
Create NullPOSProductProvider as helper

iamgabrielma committed:
Restore observeItemsInCartForCartTotal

iamgabrielma committed:
Remove formattedPrice from the POSItem protocol

iamgabrielma committed:
Update POSItem protocol

iamgabrielma committed:
Linting

iamgabrielma committed:
Move product provider creation to viewmodel

By passing the reference to the hubmenu view we avoid the issue of model recreation too often when the view recreates

jaclync commented on WooCommerce/Classes/POS/Presentation/PointOfSaleEntryPointView.swift:
```diff
@@ -1,16 +1,19 @@
 
 struct PointOfSaleEntryPointView: View {
     @StateObject private var viewModel: PointOfSaleDashboardViewModel
+    private var itemProvider: POSItemProvider
```
nit: the view probably doesn't need to own the item provider, as it's better for the view to just own the view model

iamgabrielma replied:
That's a good point, but comes back to the question of where/how should we initialize the initial `pos` state when we come from the `app`. At the moment this view just acts as a "stepping stone" for for creating the `PointOfSaleDashboardViewModel` and `PointOfSaleDashboardView`

```
PointOfSaleEntryPointView(
  itemProvider: viewModel.posItemProvider,
  currencySettings: ServiceLocator.currencySettings,
  hideAppTabBar: { isHidden in
    AppDelegate.shared.setShouldHideTabBar(isHidden)
})
```

I think this shouldn't be a view to start with, as is never actually "viewed", but is a constraint we need from the `HubMenu`, so perhaps removing the `PointOfSaleEntryPointView` entirely and creating the `PointOfSaleDashboardView` would make sense, or create that singleton for dependencies that came up in some previous conversations. I'll explore this separately 🙇

jaclync replied:
I just thought that `HubMenuViewModel` already owns `POSItemProvider`, this doesn't be owned by this POS view. Alternatively, we could also DI `POSItemProvider` to `PointOfSaleDashboardViewModel`.

jaclync commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -1,3 +1,4 @@
+import Yosemite
```
nit: maybe we can specify the enum/class/struct from Yosemite that's required in the view?

iamgabrielma replied:
Agree! I updated these here: https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/70ff262f03fbd9830a815c2af77482dbffffd91f , which has uncovered a bunch of these Yosemite imports are only needed for previews, not for the view itself, so I'll tackle removing this dependency separately here:  https://github.com/woocommerce/woocommerce-ios/issues/12917

jaclync commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -16,9 +17,9 @@ struct CartView: View {
-                ForEach(viewModel.productsInCart, id: \.product.productID) { cartProduct in
-                    ProductRowView(cartProduct: cartProduct) {
-                        viewModel.removeProductFromCart(cartProduct)
+                ForEach(viewModel.itemsInCart, id: \.item.productID) { cartItem in
```
❓since there could be multiple cart items of the same product (ID), does using the item's product ID as the `id` result in SwiftUI warning from duplicated IDs? Similar to the fix in https://github.com/woocommerce/woocommerce-ios/commit/2571a392e7dd917e2500eb44d6be1d8bce75c749

iamgabrielma replied:
Good catch, and a nice reminder to get an eye on the console!

```
ForEach<Array<CartItem>, Int64, ModifiedContent<ModifiedContent<ItemRowView, _BackgroundStyleModifier<Color>>, _PaddingLayout>>: the ID 22 occurs multiple times within the collection, this will give undefined results!
```

Resolved here: https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/9686125b628ecd2d9492e1197c66d24a1bc0c13b

jaclync commented on WooCommerce/Classes/POS/Presentation/ItemRowView.swift:
```diff
@@ -1,22 +1,23 @@
+import Yosemite
```
nit: maybe we can specify the struct/enum/class required for import? (similar for other views with Yosemite import)

iamgabrielma replied:
Yep! Updated here: https://github.com/woocommerce/woocommerce-ios/commit/70ff262f03fbd9830a815c2af77482dbffffd91f

iamgabrielma commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -1,19 +1,19 @@
-        self.siteID = ServiceLocator.stores.sessionManager.defaultSite?.siteID ?? 0
-        self.currencySettings = ServiceLocator.currencySettings
+    public init(storageManager: StorageManagerType, siteID: Int64, currencySettings: CurrencySettings ) {
+        // ServiceLocator is part of the WooCommerce binary, expose
```
Oops I committed this comment by mistake, I'll remove it before merging.

iamgabrielma replied:
Removed on https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/6a4ec649a6b2a40a07a8e735024f6f707e12be0e

jaclync commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -66,20 +67,31 @@ final class POSProductProvider {
+
+/// Null product provider that acts as preview helper for the Point of Sale feature
+///
+public final class NullPOSProductProvider: POSItemProvider {
```
nit: since this seems to be used only for SwiftUI previews within DEBUG, how about moving this to the app layer within DEBUG?

iamgabrielma replied:
Agree, I've added it to the same issue as the Yosemite dependencies for previews only. Will be tackled separately here: https://github.com/woocommerce/woocommerce-ios/issues/12917

jaclync commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenu.swift:
```diff
@@ -162,7 +162,11 @@ private extension HubMenu {
             case HubMenuViewModel.PointOfSaleEntryPoint.id:
-                PointOfSaleEntryPointView(currencySettings: ServiceLocator.currencySettings, hideAppTabBar: { isHidden in
+                // TODO:
+                // PointOfSaleEntryPointView does not need currencySettings, since these are passed through POSItemProvider
```
nit: just checking, can we already remove currencySettings here, or is it still needed for other POS use cases and we plan to remove it in the future?

iamgabrielma replied:
I believe it can be removed, I've raised the same question here: https://github.com/woocommerce/woocommerce-ios/pull/12882#discussion_r1621778958 and will tackle it on https://github.com/woocommerce/woocommerce-ios/issues/12918 just to keep things separate 👍

jaclync commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift:
```diff
@@ -70,6 +70,18 @@ final class HubMenuViewModel: ObservableObject {
+    // Is this the right place to instantiate the product provider and use property injection?
+    lazy var posItemProvider: POSItemProvider = {
+        let storageManager = ServiceLocator.storageManager
+        let siteID = ServiceLocator.stores.sessionManager.defaultSite?.siteID ?? 0
```
❓  is the siteID from ServiceLocator instead of `HubMenuViewModel.siteID` because of the issue in p1717043548835669/1716913986.223809-slack-CGPNUU63E where the `HubMenuViewModel.siteID` isn't accurate after switching stores? I think if `HubMenuViewModel.siteID` is indeed inaccurate, it's worth creating 
an issue as it might cause other issues within the Menu.

iamgabrielma replied:
That's a good question that I haven't had the chance to dig much into it yet. I wouldn't expect that those ID's are different, but they could be 🤔  specially since our `pos` dependencies are created as lazy through the lifecycle of a different component that relies its recreation directly on site switching. I've logged this separately as definitely needs more testing: https://github.com/woocommerce/woocommerce-ios/issues/12919

jaclync commented on WooCommerce/Classes/ViewRelated/Hub Menu/HubMenuViewModel.swift:
```diff
@@ -70,6 +70,18 @@ final class HubMenuViewModel: ObservableObject {
     private let posEligibilityChecker: POSEligibilityCheckerProtocol
 
+    // TODO:
+    // Is this the right place to instantiate the product provider and use property injection?
```
I think this is good for now, we can revisit how we pass the view model to the entry point view later. If possible, maybe this variable can be marked as `private(set)`.

iamgabrielma replied:
Thanks, revisiting when clarifying DI makes sense to me. Also updated access control on https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/9b988f09860e9ff963755383aa962ac9f8b53f6c

jaclync commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -1,19 +1,19 @@
-        self.storageManager = ServiceLocator.storageManager
-        self.siteID = ServiceLocator.stores.sessionManager.defaultSite?.siteID ?? 0
-        self.currencySettings = ServiceLocator.currencySettings
+    public init(storageManager: StorageManagerType, siteID: Int64, currencySettings: CurrencySettings ) {
```
super nit:
```suggestion
    public init(storageManager: StorageManagerType, siteID: Int64, currencySettings: CurrencySettings) {
```

iamgabrielma replied:
Good eye! Fixed: https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/599fc35a1337b7998bbe8bc3f4995ad938dd6b80

jaclync commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -66,20 +67,31 @@ final class POSProductProvider {
     }
 
-    static func provideProductsForPreview(currencySettings: CurrencySettings = ServiceLocator.currencySettings) -> [POSProduct] {
+    public static func provideProductsForPreview() -> [POSProduct] {
```
❓ Now that we have a protocol `POSItemProvider` and`POSItem` and this is just for SwiftUI previews, how about creating another implementation of `POSItemProvider` & `POSItem` within DEBUG in the app layer to provide mock items for previews?

iamgabrielma replied:
Agree, as the other preview-related feedback I've added this to https://github.com/woocommerce/woocommerce-ios/issues/12917 so can be tackled together.

jaclync commented on Yosemite/Yosemite/PointOfSale/POSItemProvider.swift:
```diff
@@ -0,0 +1,14 @@
+    var imageURL: URL { get }
+    var details: [String] { get }
+
+    func makeCartItem() -> CartItem
```
❓ this doesn't seem used in the PR, is this necessary? It looks like the `CartItem` is created in `PointOfSaleDashboardViewModel.addItemToCart`.

If this isn't needed now, maybe we can move `CartItem` back to the app layer, because the POSItemProvider doesn't need to know about the cart. If we need the cart item in the data layer (Yosemite for now) (possibly when syncing order), we can create another version in Yosemite for what's needed.

iamgabrielma replied:
These are not yet necessary, and doesn't seem that provider needs to know about them indeed. I've removed the additional properties and moved it back to app layer on https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/e8d12cd335c3a3836f271d1e2731aa44bf6c2db1 and https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/4754c81d66ce3627543aae8f8e74d5d53e96b768

jaclync commented on Yosemite/Yosemite/PointOfSale/POSProduct.swift:
```diff
@@ -0,0 +1,33 @@
+
+public struct POSProduct: POSItem {
```
❓ `POSProduct`'s public use seems to be for SwiftUI previews in the app layer, maybe we can create a separate implementation of `POSItem` for previews within DEBUG and name it like `POSPreviewItem`/`POSItemForPreview` (I'm not good at naming things)? In that case, `POSProduct` can probably be internal within Yosemite.

iamgabrielma replied:
Sounds good, added to the other SwiftUI preview cases that use Yosemite as dependency on https://github.com/woocommerce/woocommerce-ios/issues/12917

jaclync commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -51,11 +51,12 @@ final class POSProductProvider {
                        name: product.name,
                        price: product.price,
-                       currencySettings: currencySettings)
+                       formattedPrice: "$\(product.price)")
```
just checking, do you plan to create the price with currencySettings in this or a separate PR?

iamgabrielma replied:
I'm leaving it for a separate PR, since we plan to remove the currency settings dependency duplication on https://github.com/woocommerce/woocommerce-ios/issues/12918 . It may also be a good place to start simple adding unit test for the models and be sure currency and others are passed correctly.

jaclync reviewed (commented):
Overall LGTM, just wanted to clarify a few cases in comments marked with ❓(some might be related) !

iamgabrielma committed:
Remove internal comment

iamgabrielma committed:
Add specificity when importing Yosemite components

iamgabrielma committed:
Fix duplicate ID warning in SwiftUI

When looping through a collection using keypaths the ID is suppossed to be unique hashables to differentiate views, but if we rely on the product ID this won’t happen since multiple of the same product ID can be added to the cart. We move to use their unique ID instead.

iamgabrielma committed:
Lint

iamgabrielma committed:
Update item provider access control

iamgabrielma committed:
Remove not-yet-used model properties

iamgabrielma committed:
Move CartItem to app layer

iamgabrielma commented:
Thanks for the review and guidance @jaclync and @joshheald 🙇 

This is ready for review, there are some items that will be tackled separately so there are not that many things going on within the same PR:

- Remove Yosemite dependencies when are only needed for previews: https://github.com/woocommerce/woocommerce-ios/issues/12917
- Resolve currency settings duplication: https://github.com/woocommerce/woocommerce-ios/issues/12918
- Make `CartItem` a protocol: https://github.com/woocommerce/woocommerce-ios/issues/12920

~~There are a bunch of conflicts in the affected files due to the nature of the changes (mostly previews since we've also added the card reader service to the vm init), if we're happy with this PR I'll start to resolve these.~~ conflicts have been addressed, please note that watchOS simulator is needed (~4GB) in order to run the WooCommerce scheme.

iamgabrielma committed:
Merge branch 'trunk' into task/12879-pos-move-product-provider-to-yosemite

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/CartView.swift
#	WooCommerce/Classes/POS/Presentation/FilterView.swift
#	WooCommerce/Classes/POS/Presentation/ItemGridView.swift
#	WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift
#	WooCommerce/Classes/POS/Presentation/PointOfSaleEntryPointView.swift
#	WooCommerce/Classes/POS/Presentation/TotalsView.swift
#	WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

jaclync commented on WooCommerce/Classes/POS/Models/CartItem.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+import protocol Yosemite.POSItem
+
+public struct CartItem {
+    public let id: UUID
+    public let item: POSItem
+    public let quantity: Int
+
+    public init(id: UUID, item: POSItem, quantity: Int) {
```
nit: in the app layer, `public` is probably not needed

iamgabrielma replied:
Fixed! https://github.com/woocommerce/woocommerce-ios/pull/12882/commits/26c95046e40dca11f1cdfa89f45c5241e5b355bb

iamgabrielma committed:
Resolve merge conflicts

jaclync commented:
I started reviewing the PR, might have to pause for a few hours but I aim to finish it by the EOD.

jaclync reviewed (approved):
Thanks for making the updates and documenting the follow-up tasks clearly 💟 LGTM and tests well!

iamgabrielma committed:
Make CartItem internal
